examples/asrc_demo/src/avg_buffer_level.c

```diff
@@ -0,0 +1,63 @@
+#define DEBUG_UNIT RATE_SERVER
+#define DEBUG_PRINT_ENABLE_RATE_SERVER 0
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+#include "rtos_printf.h"
+#include "avg_buffer_level.h"
+
+void init_calc_buffer_level_state(buffer_calc_state_t *p_calc_state, int32_t window_len_log2, int32_t buffer_level_stable_threshold)
+{
+    memset(p_calc_state, 0, sizeof(buffer_calc_state_t));
+    p_calc_state->window_len_log2 = window_len_log2;
+    p_calc_state->buffer_level_stable_threshold = buffer_level_stable_threshold;
+}
+
+void calc_avg_buffer_level(buffer_calc_state_t *state, int current_level, bool reset)
+{
+
+    if(reset == true)
+    {
+        init_calc_buffer_level_state(state, state->window_len_log2, state->buffer_level_stable_threshold); // Reinitialise state
+        rtos_printf("Reset avg buffer level\n");
+        return;
+    }
+
+    state->error_accum += current_level;
+
+    state->count += 1;
+
+    if(state->count == (1<<state->window_len_log2))
+    {
+        if(state->flag_first_done == true)
+        {
+            state->flag_prev_avg_valid = true;
+        }
+        state->prev_avg_buffer_level = state->avg_buffer_level;
```

Comment:
I think `prev_avg_buffer_level` does not need to be part of state as the value is only read on line 42 so it could be a local variable on the stack